tabbycat/options/tests/management/commands/test_applypreset.py

```diff
@@ -0,0 +1,45 @@
+from unittest.mock import patch
+
+from django.core.management import call_command
+from django.core.management.base import CommandError
+from django.test import TestCase
+
+from options.management.commands.applypreset import Command
+from options.presets import PreferencesPreset
+from tournaments.models import Tournament
+
+
+class TestPreset(PreferencesPreset):
+    name         = "Test Rules"
+    description  = "Used for unit tests"
+    show_in_list = False
+
+    # Scoring
+    scoring__score_min = 70
+    scoring__score_max = 80
+
+
+class ApplyPresetTests(TestCase):
+    @patch('options.management.commands.applypreset.all_presets', return_value=[TestPreset])
+    def test_preset_options(self, mock_all_presets):
+        with self.assertRaises(CommandError) as cm:
+            Command().create_parser('', '').parse_args(['notvalidpreset'])
+            self.assertEqual(str(cm), "Error: argument preset: invalid choice: 'notvalidpreset' (choose from 'testpreset')")
```

Comment:
No, the `assertRaises` acts as a `try/catch`, and we need the `assertEqual` to have the contents of the exception, which would only be available in the `with` context.

Reply:
I don't think that's how it works? The example here definitely uses `cm` outside the context:
https://docs.python.org/3/library/unittest.html#unittest.TestCase.assertRaises

``` python
with self.assertRaises(SomeException) as cm:
    do_something()

the_exception = cm.exception
self.assertEqual(the_exception.error_code, 3)
```

Otherwise where would the equivalent `try ... except` block go? The context manager can't tell the difference between anything inside the block—it can't differentiate between the first and second lines, in particular, because the `yield` statement can only go through the whole block. So it would have to translate to
``` python
try:
    Command().create_parser('', '').parse_args(['notvalidpreset'])
    # execution never reaches this line if the above line raises an error
    self.assertEqual(str(cm), "Error: argument preset: invalid choice: 'notvalidpreset' (choose from 'testpreset')")
except CommandError as e:
    # what would it do with the error???
```
My understanding is that the `assertRaises()` context manager is equivalent to
``` python
try:
    Command().create_parser('', '').parse_args(['notvalidpreset'])
    # I think this is the wrong place...
    # self.assertEqual(str(cm), "Error: argument preset: invalid choice: 'notvalidpreset' (choose from 'testpreset')")
except CommandError as cm:
    cm.exception = e

# Now you can use cm.exception
self.assertEqual(str(cm.exception), "Error: argument preset: invalid choice: 'notvalidpreset' (choose from 'testpreset')")
```

----------------------

If you don't believe me, the easiest way to test this is to modify the expected error message to be wrong:
``` suggestion
            self.assertEqual(str(cm), "This is not the expected error message")
```
Try that, and you'll find that both tests (this and the one below) incorrectly pass.

Try unindenting this line:
```suggestion
        self.assertEqual(str(cm), "Error: argument preset: invalid choice: 'notvalidpreset' (choose from 'testpreset')")
```
and I get this:
```
AssertionError: '<unittest.case._AssertRaisesContext obje[17 chars]af0>' != "Error: argument preset: invalid choice: [39 chars]et')"
- <unittest.case._AssertRaisesContext object at 0x7fc3e3bd8af0>
+ Error: argument preset: invalid choice: 'notvalidpreset' (choose from 'testpreset')
```

I think this is because you're meant to use `cm.exception`:
```suggestion
            self.assertEqual(str(cm.exception), "Error: argument preset: invalid choice: 'notvalidpreset' (choose from 'testpreset')")
```
This passes.

And to be sure, change it to the wrong message:
```suggestion
            self.assertEqual(str(cm), "This is not the expected error message")
```
And this fails with:
```
AssertionError: "Error: argument preset: invalid choice: [39 chars]et')" != 'This is not the expected error message'
- Error: argument preset: invalid choice: 'notvalidpreset' (choose from 'testpreset')
+ This is not the expected error message
```
which is the expected failure.